### PR TITLE
Fix issue 16774: talk about constructable stylesheets in the shadow D…

### LINKED ISSUE #16774: [Web-API] Constructible stylesheets guide
FF101 minimal release work for constructible stylesheets was done in #16411. This covered release notes and new properties. 

What still need to be done is to provide an explanation of the API as "guide" material.
IN summary, it's a way to construct a CSS stylesheet programmatically so that it can be shared efficiently with shadow DOM subtrees.

This is mostly likely an addition to https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM and lots of cross linking.

The main methods/properties are:
- [CSSStyleSheet](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet) methods including its [CSSStyleSheet()](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/CSSStyleSheet), `replace`, `replaceSync`, `insertRule`, `deleteRule` and `adoptedStylesheets` in Document and ShadowRoot.

Resources
- Explainer: https://github.com/WICG/construct-stylesheets/blob/gh-pages/explainer.md
- [Constructable Stylesheets](https://web.dev/constructable-stylesheets/) (web.dev) 

- [ ] When done, mark this one dev-doc-complete https://bugzilla.mozilla.org/show_bug.cgi?id=1644102

### PR DESCRIPTION
**Pull request by wbamberg**
Fixes https://github.com/mdn/content/issues/16774.

I just rewrote the shadow DOM guide in https://github.com/mdn/content/pull/29355, and thought this would be a good time to tackle https://github.com/mdn/content/issues/16774.

Maybe it should be more extensive? But then I think it should maybe be a separate page.

### CODE REVIEW

**Comment by wbamberg** [comment]
Actually I should also update https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements and most of the examples in mdn/web-components-examples. Let me know if you want that in this PR or a different one (I would prefer different).

File: files/en-us/web/api/web_components/using_shadow_dom/index.md [hunk 1]
```diff
@@ -243,6 +250,12 @@ shadow.appendChild(span);
 
 So the encapsulation of CSS works both ways: page styles don't affect the shadow DOM tree, and shadow DOM styles don't affect elements outside the shadow DOM.
 
+You can also apply styles inside a shadow DOM by creating a {{htmlelement("style")}} element and adding it to the shadow DOM tree. However, this is inefficient if you have many shadow DOM trees in a document, because the browser may need to parse the rules within every `<style>` element.
```
Comment (estelle) [inline files/en-us/web/api/web_components/using_shadow_dom/index.md:253]: Not sure if this is accurate, but even if it is, I am not sure we should recommend against `<style>`, as it's the simplest way to create a component using `<template>`. Marinating on this.

**Comment by wbamberg** [comment]
FWIW "reset" not "reload" was intentional: reloading the page is the implementation (because it's in an iframe) but resetting the state of the example is the user-visible effect.

File: files/en-us/web/api/web_components/using_shadow_dom/index.md [hunk 2]
```diff
@@ -206,7 +206,13 @@ The page CSS does not affect nodes inside the shadow DOM:
 
 ## Applying styles inside the shadow DOM
 
-To style page elements in the shadow DOM, we can create a {{htmlelement("style")}} element and attach it to the shadow root. Rules defined in here will be scoped to the shadow DOM tree.
+To style page elements in the shadow DOM, we can:
```
Comment (estelle) [inline files/en-us/web/api/web_components/using_shadow_dom/index.md:209]:
```suggestion
To style page elements in the shadow DOM, we can create a {{htmlelement("style")}} element and attach it to the shadow root, [include a `<style>`](/en-US/docs/Web/API/Web_components/Using_templates_and_slots#creating_a_template_with_some_slots) directly within a with {{htmlelement("template")}} , or  attach a style sheet to the shadow root. The rules within are scoped to the shadow DOM tree in all cases.

The steps to attach a style sheet to the shadow root include:
```

stating there is an alternative, then doing the step.

File: files/en-us/web/api/web_components/using_shadow_dom/index.md [hunk 3]
```diff
@@ -206,7 +206,13 @@ The page CSS does not affect nodes inside the shadow DOM:
...
+
+- Create an empty {{domxref("CSSStyleSheet")}} object
+- Set its content using {{domxref("CSSStyleSheet.replace()")}} or {{domxref("CSSStyleSheet.replaceSync()")}}
+- Add it to the shadow root by assigning it to {{domxref("ShadowRoot.adoptedStyleSheets")}}
```
Comment (estelle) [inline files/en-us/web/api/web_components/using_shadow_dom/index.md:213]:
```suggestion
1. Create an empty {{domxref("CSSStyleSheet")}} object
2. Set its content using {{domxref("CSSStyleSheet.replace()")}} or {{domxref("CSSStyleSheet.replaceSync()")}}
3. Add it to the shadow root by assigning it to {{domxref("ShadowRoot.adoptedStyleSheets")}}
```
These are steps, so numbering them

File: files/en-us/web/api/web_components/using_shadow_dom/index.md [hunk 4]
```diff
@@ -243,6 +250,12 @@ shadow.appendChild(span);
 
 So the encapsulation of CSS works both ways: page styles don't affect the shadow DOM tree, and shadow DOM styles don't affect elements outside the shadow DOM.
 
+You can also apply styles inside a shadow DOM by creating a {{htmlelement("style")}} element and adding it to the shadow DOM tree. However, this is inefficient if you have many shadow DOM trees in a document, because the browser may need to parse the rules within every `<style>` element.
+
+Creating a `CSSStyleSheet` and assigning it to the shadow root using `adoptedStyleSheets` allows us to create a single stylesheet and share it among many DOM trees.
+
+For example, a component library could create a single stylesheet and then share it among all the custom elements belonging to that library.
```
Comment (estelle) [inline files/en-us/web/api/web_components/using_shadow_dom/index.md:257]:
```suggestion
Applying CSS styles inside a shadow DOM by creating a {{htmlelement("style")}} element and adding it to the shadow DOM tree may seem inefficient if you have many shadow DOM trees in a document as the browser needs to parse each occurrence, browsers are extremely efficient at parsing CSS. There are likely lower-hanging fruit with a much greater impact on performance optimization available that are more impactful places to focus on to improve performance.

While adding a `<style>` block to a `<template>` means each occurrence of the custom element may include the same style block, the template is only downloaded once, meaning duplicate bytes are not an issue. And, again, browsers are optimized for parsing CSS.

Creating a `CSSStyleSheet` and assigning it to the shadow root using `adoptedStyleSheets` allows us to create a single stylesheet and share it among many DOM trees. Similarly, the `<style>` within the `<template>` is a single style block shared among each occurrence of the custom element.

For example, a component library could include styles directly within each custom element that are element-specific and create a single global stylesheet for the library that is shared among all the custom elements within that library.
```

File: files/en-us/web/api/web_components/using_shadow_dom/index.md [hunk 5]
```diff
@@ -292,3 +305,7 @@ For more examples, illustrating different aspects of custom element implementati
 
 - [Using custom elements](/en-US/docs/Web/API/Web_components/Using_custom_elements)
 - [Using templates and slots](/en-US/docs/Web/API/Web_components/Using_templates_and_slots)
+- {{domxref("Element.attachShadow()")}}
+- {{domxref("ShadowRoot.adoptedStyleSheets")}}
+- {{domxref("CSSStyleSheet.replace()")}}
+- {{domxref("CSSStyleSheet.replaceSync()")}}
```
Comment (estelle) [inline files/en-us/web/api/web_components/using_shadow_dom/index.md:311]:
```suggestion
- {{domxref("CSSStyleSheet.replaceSync()")}}
- {{HTMLelement("template")}}
- [CSS scoping](/en-US/docs/Web/CSS/CSS_scoping) module
- {{CSSXref(":host")}}
- {{CSSXref(":host_function", ":host()")}}
- {{CSSXref(":host-context", ":host-context()")}}
- {{CSSXref("::slotted", "::slotted()")}}
- [CSS shadow parts](/en-US/docs/Web/CSS/CSS_shadow_parts) module
- {{CSSXref("::part")}}

```

**Review by estelle — COMMENTED** [review]
i made some suggests that are kind of more guidance as to what i think we should include, not necessary the words that need to be used.

I also directly committed: I changed the "reset" button so it reads "reload". The example in the page (not your edits) was taking 3 seconds to "reset", and I stared at the JS to figure out why the event handler was having issues before realizing it was a reload not a reset.

**Comment by estelle** [comment]
> FWIW "reset" not "reload" was intentional: reloading the page is the implementation (because it's in an iframe) but resetting the state of the example is the user-visible effect.

Didn't see your comment until after my comment. Still think we should use reload. maybe "reload iframe". The "reset" taking 3 seconds made me wonder why a Shadow Dom edit was slow. Maybe it's b/c it's a temp page. I should likely test live.....

**Comment by wbamberg** [comment]
> > FWIW "reset" not "reload" was intentional: reloading the page is the implementation (because it's in an iframe) but resetting the state of the example is the user-visible effect.
> 
> Didn't see your comment until after my comment. Still think we should use reload. maybe "reload iframe". The "reset" taking 3 seconds made me wonder why a Shadow Dom edit was slow. Maybe it's b/c it's a temp page. I should likely test live.....

For me it is almost instant, perhaps my Internet is too fast! If it is taking that long then location.reload() is not a good way to do this, and we should reset "manually" instead. I use this a lot in live samples to reset, because it's just a really concise way to do it.

I think "reload iframe" is worse. The fact that the example is in an iframe is an implementation detail that's totally hidden from readers.

**Comment by wbamberg** [comment]
I just pushed https://github.com/mdn/content/pull/29484/commits/0d05c890cad3acb82b718644988abf7eaac7e739, which tries to expand on different ways to apply styles to the shadow DOM, with different sections for two approaches, and a bit comparing them that's based on https://github.com/mdn/content/pull/29267#issuecomment-1751596996.

**Review by estelle — APPROVED** [review]
Looks great. Really easy to follow

**Pull request merged**